python-checks/src/main/java/org/sonar/python/checks/cdk/PublicNetworkAccessToCloudResourcesCheck.java

```diff
@@ -33,5 +62,109 @@ protected void registerFqnConsumer() {
+      .flatMap(dictionary -> getDictionaryPair(ctx, dictionary, SUBNET_TYPE))
+      .flatMap(element -> element.value.getExpression(isFqn(SENSITIVE_SUBNET)))
+      .ifPresent(expression -> raiseIssueOnParent(ctx, expression, Tree.Kind.REGULAR_ARGUMENT));
+  }
+
+  /* Methods to work on aws_cdk.aws_ec2.CfnInstance objects */
+  // Check this expression : 'aws_cdk.aws_ec2.CfnInstance(network_interfaces=[*Sensitive network interface*])'
+  private static void checkCfnInstance(SubscriptionContext ctx, CallExpression callExpression) {
+    getArgument(ctx, callExpression, "network_interfaces")
+      .flatMap(flow -> flow.getExpression(isListLiteral())).map(ListLiteral.class::cast)
+      .map(listLiteral -> listLiteral.elements().expressions())
+      .orElse(Collections.emptyList())
+      .forEach(expression -> {
+        checkNetworkInterfacesCallExpression(ctx, expression);
+        checkNetworkInterfacesDictionary(ctx, expression);
+      });
+  }
+
+  // Check the *Sensitive network interface* as a specific CallExpression :
+  // 'aws_cdk.aws_ec2.CfnInstance.NetworkInterfaceProperty(associate_public_ip_address=True, subnet_id=*sensitive_or_no_subnet*)'
+  private static void checkNetworkInterfacesCallExpression(SubscriptionContext ctx, Expression expression) {
+    Optional.of(expression)
+      .filter(isCallExpression()).map(CallExpression.class::cast)
+      .filter(isFqn("aws_cdk.aws_ec2.CfnInstance.NetworkInterfaceProperty"))
+      .ifPresent(call -> checkSensitiveOptionWithoutCompliantSubnetDefined(ctx, call));
+  }
+
+  private static void checkSensitiveOptionWithoutCompliantSubnetDefined(SubscriptionContext ctx, CallExpression callExpression) {
+    Optional<CdkUtils.ExpressionFlow> associatedPublicIpAddress = getArgument(ctx, callExpression, ASSOCIATE_PUBLIC_IP_ADDRESS);
+    Optional<CdkUtils.ExpressionFlow> subnetId = getArgument(ctx, callExpression, "subnet_id");
+
+    if (associatedPublicIpAddress.filter(flow -> flow.hasExpression(isTrue())).isPresent()
+      && subnetId.filter(PublicNetworkAccessToCloudResourcesCheck::hasPrivateSubnetDefined).isEmpty()) {
+      associatedPublicIpAddress.get().addIssue(ERROR_MESSAGE);
+    }
+  }
+
+  // Check the *Sensitive network interface* as a specific DictionaryLiteral : '{"associate_public_ip_address" : True, "subnet_id" : *sensitive_or_no_subnet*}'
+  private static void checkNetworkInterfacesDictionary(SubscriptionContext ctx, Expression expression) {
+    List<CdkUtils.ResolvedKeyValuePair> elements = getDictionary(expression)
+      .map(dictionary -> resolveDictionary(ctx, dictionary))
+      .orElse(Collections.emptyList());
+
+    Optional<CdkUtils.ResolvedKeyValuePair> associatedPublicIpAddress = elements.stream().filter(el -> el.key.hasExpression(isString(ASSOCIATE_PUBLIC_IP_ADDRESS))).findFirst();
+    Optional<CdkUtils.ExpressionFlow> subnetId = elements.stream().filter(element -> element.key.hasExpression(isString("subnet_id"))).map(el -> el.value).findFirst();
```

Comment:
we're iterating twice over `elements` , what about using a plain for loop and check both conditions inside the body?